agent/app/agent_unix.go

```diff
@@ -171,16 +171,20 @@ func (agent *ecsAgent) verifyCNIPluginsCapabilities() error {
 // notifications from the monitor
 func (agent *ecsAgent) startUdevWatcher(state dockerstate.TaskEngineState, stateChangeEvents chan<- statechange.Event) error {
 	seelog.Debug("Setting up ENI Watcher")
-	udevMonitor, err := udevwrapper.New()
-	if err != nil {
-		return errors.Wrapf(err, "unable to create udev monitor")
-	}
-	// Create Watcher
-	eniWatcher := watcher.New(agent.ctx, agent.mac, udevMonitor, state, stateChangeEvents)
-	if err := eniWatcher.Init(); err != nil {
-		return errors.Wrapf(err, "unable to initialize eni watcher")
+	if agent.udevMonitor == nil {
+		monitor, err := udevwrapper.New()
+		if err != nil {
+			return errors.Wrapf(err, "unable to create udev monitor")
+		}
+		agent.udevMonitor = monitor
+
+		// Create Watcher
+		eniWatcher := watcher.New(agent.ctx, agent.mac, agent.udevMonitor, state, stateChangeEvents)
+		if err := eniWatcher.Init(); err != nil {
+			return errors.Wrapf(err, "unable to initialize eni watcher")
+		}
+		go eniWatcher.Start()
```

Comment:
i think this part should be outside of the if block so that the eni watcher can always be started

Reply:
I had considered it but thought with previous implementation, one watcher was guaranteed given that it returns error if monitor couldn't be started few lines above. Wouldn't this result in behavior change?